junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/extension/TimeoutConfiguration.java

```diff
@@ -0,0 +1,106 @@
+
+class TimeoutConfiguration {
+
+	private static final Logger logger = LoggerFactory.getLogger(TimeoutConfiguration.class);
+
+	private final TimeoutDurationParser parser = new TimeoutDurationParser();
+	private final Map<String, Optional<TimeoutDuration>> cache = new ConcurrentHashMap<>();
+	private ExtensionContext extensionContext;
+
+	TimeoutConfiguration(ExtensionContext extensionContext) {
+		this.extensionContext = extensionContext;
+	}
+
+	Optional<TimeoutDuration> getDefaultTestMethodTimeout() {
+		return parseOrDefault(DEFAULT_TEST_METHOD_TIMEOUT_PROPERTY_NAME, this::getDefaultTestableMethodTimeout);
+	}
+
+	Optional<TimeoutDuration> getDefaultTestTemplateMethodTimeout() {
+		return parseOrDefault(DEFAULT_TEST_TEMPLATE_METHOD_TIMEOUT_PROPERTY_NAME,
+			this::getDefaultTestableMethodTimeout);
+	}
+
+	Optional<TimeoutDuration> getDefaultTestFactoryMethodTimeout() {
+		return parseOrDefault(DEFAULT_TEST_FACTORY_METHOD_TIMEOUT_PROPERTY_NAME, this::getDefaultTestableMethodTimeout);
+	}
+
+	Optional<TimeoutDuration> getDefaultBeforeAllMethodTimeout() {
+		return parseOrDefault(DEFAULT_BEFORE_ALL_METHOD_TIMEOUT_PROPERTY_NAME, this::getDefaultLifecycleMethodTimeout);
+	}
+
+	Optional<TimeoutDuration> getDefaultBeforeEachMethodTimeout() {
+		return parseOrDefault(DEFAULT_BEFORE_EACH_METHOD_TIMEOUT_PROPERTY_NAME, this::getDefaultLifecycleMethodTimeout);
+	}
+
+	Optional<TimeoutDuration> getDefaultAfterEachMethodTimeout() {
+		return parseOrDefault(DEFAULT_AFTER_EACH_METHOD_TIMEOUT_PROPERTY_NAME, this::getDefaultLifecycleMethodTimeout);
+	}
+
+	Optional<TimeoutDuration> getDefaultAfterAllMethodTimeout() {
+		return parseOrDefault(DEFAULT_AFTER_ALL_METHOD_TIMEOUT_PROPERTY_NAME, this::getDefaultLifecycleMethodTimeout);
+	}
+
+	private Optional<TimeoutDuration> getDefaultTestableMethodTimeout() {
+		return parseOrDefault(DEFAULT_TESTABLE_METHOD_TIMEOUT_PROPERTY_NAME, this::getDefaultTimeout);
+	}
+
+	private Optional<TimeoutDuration> getDefaultLifecycleMethodTimeout() {
+		return parseOrDefault(DEFAULT_LIFECYCLE_METHOD_TIMEOUT_PROPERTY_NAME, this::getDefaultTimeout);
+	}
+
+	private Optional<TimeoutDuration> getDefaultTimeout() {
+		return parseTimeoutDuration(DEFAULT_TIMEOUT_PROPERTY_NAME);
+	}
+
+	private Optional<TimeoutDuration> parseOrDefault(String propertyName,
+			Supplier<Optional<TimeoutDuration>> defaultSupplier) {
+		Optional<TimeoutDuration> timeoutConfiguration = parseTimeoutDuration(propertyName);
+		return timeoutConfiguration.isPresent() ? timeoutConfiguration : defaultSupplier.get();
+	}
+
+	private Optional<TimeoutDuration> parseTimeoutDuration(String propertyName) {
+		return cache.computeIfAbsent(propertyName, key -> extensionContext.getConfigurationParameter(key).map(value -> {
+			try {
+				return parser.parse(value);
+			}
+			catch (Exception e) {
+				logger.warn(e,
+					() -> String.format("Ignored invalid timeout '%s' set via the '%s' configuration parameter.", value,
+						key));
+				return null;
```

Comment:
That violates the contract for `ConcurrentHashMap`:

> Like Hashtable but unlike HashMap, this class does not allow null to be used as a key or value.

Or am I missing something?

Reply:
Actually, I did have a test for this and

> If the mapping function returns {@code null}, no mapping is recorded.

Reply:
Ahhh... good to know. Thanks!